scipy/stats/_multivariate.py

```diff
@@ -5067,3 +5068,257 @@ def rvs(self, size=1, random_state=None):
         method.__doc__, mhg_docdict_noparams)
     method.__doc__ = doccer.docformat(method.__doc__,
                                       mhg_docdict_params)
+
+
+class conditional_table_gen(multi_rv_generic):
+    r"""Distribution of random tables with given marginals.
+
+    This is the distribution of random tables with given row and column vector
+    sums. This distribution represents the set of random tables under the null
+    hypothesis that rows and columns are independent. It is used in hypothesis
+    tests of independence.
+
+    Because of assumed independence, the expected frequency of each table
+    element can be computed from the row and column sums, so that the
+    distribution is completely determined by these two vectors.
+
+    Methods
+    -------
+    mean(row, col)
+        Mean table.
+    rvs(row, col, size=1, random_state=None)
+        Draw random tables with given row and column vector sums.
+
+    Parameters
+    ----------%(_doc_callparams)s
+    %(_doc_random_state)s
+
+    Notes
+    -----
+    %(_doc_callparams_note)s
+
+    Random elements from the distribution are generated either with a shuffling
+    algorithm or with Patefield's algorithm [1]_. The shuffling algorithm has
+    O(N) time and space complexity, where N is the total sum of entries in the
+    table. Patefield's algorithm has O(K x log(N)) time complexity, where K is
+    the number of cells in the table and requires only a small constant work
+    space.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.stats import conditional_table
+
+    >>> row = [1, 5]
+    >>> col = [2, 3, 1]
+    >>> conditional_table.mean(row, col)
+    array([[0.33333333, 0.5       , 0.16666667],
+           [1.66666667, 2.5       , 0.83333333]])
+
+    Alternatively, the object may be called (as a function) to fix the row
+    and column vector sums, returning a "frozen" distribution.
+
+    >>> dist = conditional_table(row, col)
+    >>> dist.rvs(size=1, random_state=123)
+    array([[[1., 0., 0.],
+            [1., 3., 1.]]])
+
+    .. versionadded: 1.9.3
+
+    References
+    ----------
+    .. [1] Patefield's, AS 159 Appl. Statist. (1981) vol. 30, no. 1.
+    """
+
+    def __init__(self, seed=None):
+        super().__init__(seed)
+
+    def __call__(self, row, col, seed=None):
+        """Create a frozen distribution of tables with fixed column and row marginals.
+
+        See `conditional_table_frozen` for more information.
+        """
+        return conditional_table_frozen(row, col, seed)
+
+    def _process_parameters(self, row, col):
```

Comment:
Done

Reply:
The tests you added look great. There still needs to be a check for the integrality of `row`, `col`, and `x` (where applicable) and corresponding tests.

Reply:
I am not sure how to apply such a check. Checking the array's dtype is easy and fast, but that seems restrictive. I think we should also accept dtypes float32 and float64 if the numbers are integral. Checking whether `x` consists entirely of integral numbers is quite expensive, however.

Can we avoid such a check, if the functions do approximately the right thing if a user passes non-integer values? The pmf is strictly only correct for integral observations, but the implementation accepts non-integer values because we replace the factorial with the gamma function anyway.